examples/hif2a/fit_to_multiple_rbfes.py

```diff
@@ -120,9 +125,36 @@ def _blew_up(results: List[SimulationResult]) -> bool:
     # TODO: adjust this threshold a bit, move reliability calculations into fe/estimator.py or fe/model.py
     return np.isnan(du_dls).any() or (du_dls.std(1).max() > 1000)
 
-def loss_fxn(model, ff_params, mol_a, mol_b, core, label_ddG):
-    pred_ddG, stage_results = model.predict(ff_params, mol_a, mol_b, core)
-    return pseudo_huber_loss(pred_ddG - label_ddG), stage_results
+# def loss_fxn(model, ff_params, mol_a, mol_b, core, label_ddG):
+#     pred_ddG, stage_results = model.predict(ff_params, mol_a, mol_b, core)
+#     return pseudo_huber_loss(pred_ddG - label_ddG), stage_results
+
+def loss_fxn(ff_params, batch: List[Tuple[RelativeFreeEnergy, RBFEModel]]):
+    index = RBFETransformIndex()
+    index.build([edge[0] for edge in batch])
+    indices = []
+    all_results = []
+    preds = []
+    for rfe, model in batch:
+        indices.append(list(index.get_transform_indices(rfe)))
+        pred_ddG, stage_results = model.predict(ff_params, rfe.mol_a, rfe.mol_b, rfe.core)
+        all_results.extend(list(stage_results))
+        preds.append(pred_ddG)
+    # If the edges in the batch are within a cycle, correct the ddGs
+    indices = jnp.asarray(indices)
+    ind_l, ind_r = indices.T
+    try:
+        layer = construct_mle_layer(len(index), indices)
+        corrected_dg = layer(jnp.asarray(preds))
```

Comment:
Perhaps we should make this be the identity function if the edges are disconnected?

Reply:
Doing this could be more convenient, would just require changing the output type of `layer`.

Note that the input and output shapes of `layer` are not the same. (As input it can accept arrays of as many `simulated_rbfes` and `simulated_abfes` as are available (which may be `>= n_nodes`), and as output it always returns an array of `n_nodes` absolute dgs.)

At this check, https://github.com/proteneer/timemachine/blob/48623960f6dd114e0d2b05b33f55b9278b3074de/fe/cycles.py#L110-L113 we currently raise an assertion error. If instead of raising an error we wanted to return a function that passed its arguments through unmodified, then I think the function would need to return arrays of `corrected_ddgs` and `corrected_dgs` of the same shapes as the input `simulated_rbfes` and `simulated_abfes`.

(I think I would prefer to call `construct_mle_layer` only on connected comparison networks and keep the output type as-is.)

Reply:
I am happy with this as it is, just remember the non-differentiable version supporting this case so wasn't sure if we wanted to unify it.

Reply:
Follow-up: Do we generally want to define the loss in terms of [edges](https://github.com/proteneer/timemachine/blob/48623960f6dd114e0d2b05b33f55b9278b3074de/tests/test_cycle_closure.py#L67-L77), or in terms of [nodes](https://github.com/proteneer/timemachine/blob/48623960f6dd114e0d2b05b33f55b9278b3074de/tests/test_cycle_closure.py#L110-L116)?

Reply:
Thats a good question, and something I don't have a good answer for. It seems like `nodes` would provide better generalizability (at least to ABFE), and that is what our actual underlying data is.

Reply:
If we're running cycle closure, we should probably define it in terms of dGs (nodes).

Reply:
Is it reasonable to train on ddG for non-cyclic edges and training on dG for cyclic edges?

Reply:
If the edges are non-cyclic, you can trivially induce ddGs to dGs. But from a practical standpoint it should be equivalent.

Reply:
How would I induce the dGs in this case?

Reply:
>If the edges are non-cyclic, you can trivially induce ddGs to dGs.

@proteneer : Could you clarify if this comment applies also to a _disconnected_ comparison network, or only to a connected comparison network that contains no cycles?

I'm not sure how to handle the case that the comparison network is disconnected.

(If disconnected, would expect to run into issues illustrated by the [example in this comment](https://github.com/proteneer/timemachine/pull/399#discussion_r618787870).)

Reply:
I think there are two aspects

1) Disconnected Components, which we can decompose into individual independent components. 
2) A given component may or may not have cycles. 

Regardless of whether or not the component has cycles, the cycle closure correction (in the absence of experimental dGs), should still be able to induce predicted dGs, shifted by some arbitrary amount (or picking an arbitrary node as zero). You can't really compare the these predict dGs without a bias (or some sort of shifting to the mean against the label dGs). You can add a constant bias on these predicted dGs as a free parameter for fitting. If there's multiple disconnected components, then *each* cycle will necessarily need its only shifted bias (which can be terrible as it adds more parameters to the model). 

If you don't want to deal with the bias corrections and training said biases, then just compare against the cycle closure corrected ddGs. But either way, the presence of a cycle vs no cycle should not affect the choice of predicting dG vs ddG, probably best to be consistent through out.